net: lwm2m: Fix offset for post-write callback

DaveLacerteHQ opened the pull request:
The offset in post-write callback is incorrect for TLV format as the first data block contain TLV header. This fix is replacing
reamaning for offset in opaque context struct and passing it to post-write and validation callback.

The fix also corrects the offset in pull context as it was pointing at the end of the block processed data and not on the block context prior to processing.

Signed-off-by: Dave Lacerte <lacerte.dave@hydroquebec.com>

DaveLacerteHQ commented:
For Pull context :
Before the fix, the first write callback :
- The First block : BLOCK RECEIVED: offset:512 len:512 last_block:0 total_size:703179
- The Last block : BLOCK RECEIVED: offset:703179 len:203 last_block:1 total_size:703179
After the fix : 
- The First block : BLOCK RECEIVED: offset:0 len:512 last_block:0 total_size:703179
- The Last block : BLOCK RECEIVED: offset:702976 len:203 last_block:1 total_size:703179

SeppoTakalo reviewed (changes requested):
@DaveLacerteHQ How can you get blocks that are not a full block?

I think this is against the CoAP spec

https://datatracker.ietf.org/doc/html/rfc7959
Section 2
> In most cases, all blocks being transferred for a body (except for
   the last one) will be of the same size.
  
 Section 2.2
 > [2.2](https://datatracker.ietf.org/doc/html/rfc7959#section-2.2).  Structure of a Block Option
>   Three items of information may need to be transferred in a Block
   (Block1 or Block2) option:
>   o  the size of the block (SZX);
>   o  whether more blocks are following (M);
>   o  the relative number of the block (NUM) within a sequence of blocks
      with the given size.
   ...
>    The option value divided by sixteen (the NUM field) is the sequence
   number of the block currently being transferred, starting from zero.
   The current transfer is, therefore, about the "size" bytes starting
   at byte "NUM << (SZX + 4)".

>   Implementation note:  As an implementation convenience, "(val & ~0xF)
      << (val & 7)", i.e., the option value with the last 4 bits masked
      out, shifted to the left by the value of SZX, gives the byte
      position of the first byte of the block being transferred.
   
   
   Section 2.3
  > The block size implied by SZX MUST match the size of the
         payload in bytes, if the M bit is set.  (SZX does not govern
         the payload size if M is unset).

So given the spec, if you have a server that send 504 bytes of data on a first block, it is violating the spec. Only the last block can have less than a full block.

DaveLacerteHQ commented:
We are using Leshan Server 2.0.0 M10 as lwm2m server.

In the function lwm2m_engine_get_opaque_more, the first block received get the remaining as 504 (the second get 512). (remaining = in->in_cpkt->max_len - in->offset;)

The first block have an offset of 38 and a max_len of 542 (542-38 = 504) and the second one is getting an offset of 26 and a max_len of 538 (538-26 = 512).

Does the first COAP block have something more like an header inside the payload ?

SeppoTakalo commented:
I see.

I was able to replicate this on `native_sim` running against local Leshan instance

```
[00:00:12.660,000] <dbg> net_lwm2m_obj_firmware: lwm2m_firmware_set_update_state_inst: Update state = 1
[00:00:12.660,000] <inf> app_fw_update: FIRMWARE: BLOCK RECEIVED: offset:0 len:64 last_block:0
[00:00:12.660,000] <inf> app_fw_update: FIRMWARE: BLOCK RECEIVED: offset:64 len:64 last_block:0
[00:00:12.660,000] <inf> app_fw_update: FIRMWARE: BLOCK RECEIVED: offset:128 len:64 last_block:0
[00:00:12.660,000] <inf> app_fw_update: FIRMWARE: BLOCK RECEIVED: offset:192 len:64 last_block:0
[00:00:12.660,000] <inf> app_fw_update: FIRMWARE: BLOCK RECEIVED: offset:256 len:64 last_block:0
[00:00:12.660,000] <inf> app_fw_update: FIRMWARE: BLOCK RECEIVED: offset:320 len:64 last_block:0
[00:00:12.660,000] <inf> app_fw_update: FIRMWARE: BLOCK RECEIVED: offset:384 len:64 last_block:0
[00:00:12.660,000] <inf> app_fw_update: FIRMWARE: BLOCK RECEIVED: offset:448 len:60 last_block:0
[00:00:12.660,000] <dbg> net_lwm2m_client_app: socket_state: LWM2M_SOCKET_STATE_LAST
[00:00:12.720,000] <dbg> net_lwm2m_message_handling: lwm2m_engine_default_content_format: No accept option given. Assume OMA TLV.
[00:00:12.720,000] <dbg> net_lwm2m_message_handling: lwm2m_write_handler: BLOCK1: total:4100 current:512 last:0
[00:00:12.720,000] <inf> app_fw_update: FIRMWARE: BLOCK RECEIVED: offset:512 len:64 last_block:0
[00:00:12.720,000] <inf> app_fw_update: FIRMWARE: BLOCK RECEIVED: offset:576 len:64 last_block:0
[00:00:12.720,000] <inf> app_fw_update: FIRMWARE: BLOCK RECEIVED: offset:640 len:64 last_block:0
```

TLV is the only format that our block-wise write is working. But the offset calculation seem to assume opaque format.
Leshan seem to be doing the transfer correctly.

SeppoTakalo commented:
I meant only content-format apart from sending an opaque.

All other formats should reject this.

SeppoTakalo reviewed (changes requested):
As the issue seem correct, this clearly needs a fix.

However, it feels a bit wrong to calculate an offset from assumed remaining data length. I feel like there might be bugs lurking around the corner.
Not all CoAP block-wise transfers provide a SIZE1 option, so the remaining might be incorrect.

I have another proposal:
Can you convert the `remaining` field to be an offset instead:

```c
struct lwm2m_opaque_context {
	size_t len;
	size_t remaining;
};
```

This is different from CoAP block-offset, because those track the offset within a payload, but `struct lwm2m_opaque_context` should track the offset within different writes to same resource.

There should be only few places to touch, maybe like `lwm2m_rw_opaque.c` and `lwm2m_rw_oma_tlv.c`. Then all `validate_cb` and `post_write_cb`  should use the offset field from `lwm2m_opaque_context`

Commit:
net: lwm2m: Fix offset for post-write callback

The offset in post-write callback is incorrect for TLV format as
the first data block contain TLV header. This fix is replacing
reamaning for offset in opaque context struct and passing it to
post-write and validation callback.

The fix also corrects the offset in pull context as it was pointing
at the end of the block processed data and not on the block context
prior to processing.

Signed-off-by: Dave Lacerte <lacerte.dave@hydroquebec.com>

DaveLacerteHQ commented:
I have made the change you suggested but it needed to change more things (lwm2m_rw_senml_json.c, lwm2m_rw_senml_cbor, lwm2m_rw_cbor).  Tests fails at net_content_raw_cbor_test_get_opaque: (ret not equal to strlen(long_opaque)). I you have suggestion, else I will try to have some time to work on it tomorrow.

SeppoTakalo commented:
@DaveLacerteHQ Here is how to fix your commit.

Basically just:
* Only increment the `block_ctx.offset` during `lwm2m_write_handler_opaque()`. Do not touch this value within content formats.
* In content format, use `offset == 0` to detect that you are in the first block within the opaque resource. Not `len == 0`, even though it might work, I wonder if opaque content might already have `len` set from SIZE1 CoAP option.
* Minor bugfix to `lwm2m_write_handler_opaque()` which returned from write before storing the opaque context back to msg->block_ctx.
* Minor bugfix to `lwm2m_write_handler_opaque()` which returned which returned size of a complete block, instead of amount written

```diff
  diff --git a/subsys/net/lib/lwm2m/lwm2m_message_handling.c b/subsys/net/lib/lwm2m/lwm2m_message_handling.c
  index 4af668b42e5..89e72836763 100644
  --- a/subsys/net/lib/lwm2m/lwm2m_message_handling.c
  +++ b/subsys/net/lib/lwm2m/lwm2m_message_handling.c
  @@ -998,6 +998,7 @@ static int lwm2m_write_handler_opaque(struct lwm2m_engine_obj_inst *obj_inst,
          struct lwm2m_opaque_context opaque_ctx = {0};
          void *write_buf;
          size_t write_buf_len;
  +       int written = 0;
   
          if (msg->in.block_ctx != NULL) {
                  last_block = msg->in.block_ctx->last_block;
  @@ -1025,7 +1026,7 @@ static int lwm2m_write_handler_opaque(struct lwm2m_engine_obj_inst *obj_inst,
                  len = engine_get_opaque(&msg->in, write_buf, MIN(data_len, write_buf_len),
                                          &opaque_ctx, &last_pkt_block);
                  if (len <= 0) {
  -                       return len;
  +                       break;
                  }
   
   #if CONFIG_LWM2M_ENGINE_VALIDATION_BUFFER_SIZE > 0
  @@ -1055,15 +1056,15 @@ static int lwm2m_write_handler_opaque(struct lwm2m_engine_obj_inst *obj_inst,
                  if (msg->in.block_ctx && !last_pkt_block) {
                          msg->in.block_ctx->ctx.current += len;
                  }
  +               opaque_ctx.offset += len;
  +               written += len;
          }
   
          if (msg->in.block_ctx != NULL) {
                  msg->in.block_ctx->opaque = opaque_ctx;
          }
   
  -       opaque_ctx.offset += opaque_ctx.len;
  -
  -       return opaque_ctx.len;
  +       return (len < 0 ? len : written);
   }
   /* This function is exposed for the content format writers */
   int lwm2m_write_handler(struct lwm2m_engine_obj_inst *obj_inst, struct lwm2m_engine_res *res,
  diff --git a/subsys/net/lib/lwm2m/lwm2m_registry.c b/subsys/net/lib/lwm2m/lwm2m_registry.c
  index 04931c38ac0..372b57264c3 100644
  --- a/subsys/net/lib/lwm2m/lwm2m_registry.c
  +++ b/subsys/net/lib/lwm2m/lwm2m_registry.c
  @@ -1100,9 +1100,8 @@ size_t lwm2m_engine_get_opaque_more(struct lwm2m_input_context *in, uint8_t *buf
                  in_len = remaining;
          }
   
  -       opaque->offset += in_len;
          remaining -= in_len;
  -       if (opaque->offset >= opaque->len || remaining == 0) {
  +       if (opaque->offset + in_len >= opaque->len) {
                  *last_block = true;
          }
   
  diff --git a/subsys/net/lib/lwm2m/lwm2m_rw_cbor.c b/subsys/net/lib/lwm2m/lwm2m_rw_cbor.c
  index 8dc2649694a..1966f3bd930 100644
  --- a/subsys/net/lib/lwm2m/lwm2m_rw_cbor.c
  +++ b/subsys/net/lib/lwm2m/lwm2m_rw_cbor.c
  @@ -426,7 +426,7 @@ static int get_opaque(struct lwm2m_input_context *in, uint8_t *value, size_t buf
          ZCBOR_STATE_D(states, 1, ICTX_BUF_R_PTR(in), ICTX_BUF_R_LEFT_SZ(in),  1, 0);
   
          /* Get the CBOR header only on first read. */
  -       if (opaque->len == 0) {
  +       if (opaque->offset == 0) {
                  ret = zcbor_bstr_start_decode_fragment(states, &hndl);
   
                  if (!ret) {
  diff --git a/subsys/net/lib/lwm2m/lwm2m_rw_oma_tlv.c b/subsys/net/lib/lwm2m/lwm2m_rw_oma_tlv.c
  index baaa37c14e8..c93671612b3 100644
  --- a/subsys/net/lib/lwm2m/lwm2m_rw_oma_tlv.c
  +++ b/subsys/net/lib/lwm2m/lwm2m_rw_oma_tlv.c
  @@ -745,7 +745,7 @@ static int get_opaque(struct lwm2m_input_context *in, uint8_t *value,
          int size;
   
          /* Get the TLV header only on first read. */
  -       if (opaque->len == 0) {
  +       if (opaque->offset == 0) {
                  size = oma_tlv_get(&tlv, in, false);
                  if (size < 0) {
                          return size;
  diff --git a/subsys/net/lib/lwm2m/lwm2m_rw_opaque.c b/subsys/net/lib/lwm2m/lwm2m_rw_opaque.c
  index 6c90aca92f7..505d7113e4b 100644
  --- a/subsys/net/lib/lwm2m/lwm2m_rw_opaque.c
  +++ b/subsys/net/lib/lwm2m/lwm2m_rw_opaque.c
  @@ -57,7 +57,7 @@ static int get_opaque(struct lwm2m_input_context *in, uint8_t *value,
   {
          uint16_t in_len;
   
  -       if (opaque->len == 0) {
  +       if (opaque->offset == 0) {
                  coap_packet_get_payload(in->in_cpkt, &in_len);
   
                  if (in_len == 0) {
  @@ -74,6 +74,13 @@ static int get_opaque(struct lwm2m_input_context *in, uint8_t *value,
                                  opaque->len = in->block_ctx->ctx.total_size;
                          }
   
  +                       if (opaque->len == 0) {
  +                               /* No size1 option provided, use current
  +                                * payload size. This will reset on next packet
  +                                * received.
  +                                */
  +                               opaque->len = in_len;
  +                       }
                  } else {
                          opaque->len = in_len;
                  }
  diff --git a/subsys/net/lib/lwm2m/lwm2m_rw_senml_json.c b/subsys/net/lib/lwm2m/lwm2m_rw_senml_json.c
  index d256a70a916..957da6174ee 100644
  --- a/subsys/net/lib/lwm2m/lwm2m_rw_senml_json.c
  +++ b/subsys/net/lib/lwm2m/lwm2m_rw_senml_json.c
  @@ -1272,8 +1272,7 @@ static int get_opaque(struct lwm2m_input_context *in, uint8_t *value, size_t buf
                  in_len = val_opaque->length;
          }
   
  -       opaque->offset += in_len;
  -       if (opaque->offset >= opaque->len) {
  +       if (opaque->offset + in_len >= opaque->len) {
                  *last_block = true;
          }
          /* Copy data to buffer */

```

SeppoTakalo commented:
One more change. Now the code works correctly for opaque resources, that have `post_write_cb()` set.

But if that is not set, block-wise transfer will set the resource length to same as written bytes from last block. That break some interoperability tests that I have written.

So here is a fit:
```.diff
  diff --git a/subsys/net/lib/lwm2m/lwm2m_message_handling.c b/subsys/net/lib/lwm2m/lwm2m_message_handling.c
  index 89e72836763..56da33fb3a1 100644
  --- a/subsys/net/lib/lwm2m/lwm2m_message_handling.c
  +++ b/subsys/net/lib/lwm2m/lwm2m_message_handling.c
  @@ -1105,7 +1105,7 @@ int lwm2m_write_handler(struct lwm2m_engine_obj_inst *obj_inst, struct lwm2m_eng
          if (msg->in.block_ctx != NULL) {
                  /* Get block_ctx for total_size (might be zero) */
                  total_size = msg->in.block_ctx->ctx.total_size;
  -               offset = msg->in.block_ctx->ctx.current;
  +               offset = msg->in.block_ctx->opaque.offset;
   
                  LOG_DBG("BLOCK1: total:%zu current:%zu"
                          " last:%u",
  @@ -1296,6 +1296,10 @@ int lwm2m_write_handler(struct lwm2m_engine_obj_inst *obj_inst, struct lwm2m_eng
                  }
          }
   
  +       if (!res->post_write_cb) {
  +               len += offset;
  +       }
  +
          res_inst->data_len = len;
   
          if (LWM2M_HAS_PERM(obj_field, LWM2M_PERM_R)) {
```

SeppoTakalo commented:
```
DEBUG   - run status: native_sim/native/host/tests/net/lib/lwm2m/interop/net.lwm2m.interop passed
INFO    - 1/1 native_sim/native         tests/net/lib/lwm2m/interop/net.lwm2m.interop      PASSED (native 77.936s <host>)
INFO    -                                    net.lwm2m.interop.test_blockwise_1                                          PASSED      
INFO    -                                    net.lwm2m.interop.test_blockwise_2                                          PASSED      
INFO    -                                    net.lwm2m.interop.test_blockwise_3                                          PASSED      
INFO    -                                    net.lwm2m.interop.test_blockwise_4                                          PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_1                             PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_4                             PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_5                             PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_6                             PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_102                           PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_103                           PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_104                           PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_108                           PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_201                           PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_203                           PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_204                           PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_205                           PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_211                           PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_212                           PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_215                           PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_220                           PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_221                           PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_222                           PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_223                           PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_224                           PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_225                           PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_226                           PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_227                           PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_228                           PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_229                           PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_230                           PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_231                           PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_232                           PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_233                           PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_234                           PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_235                           PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_236                           PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_237                           PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_241                           PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_256                           PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_257                           PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_260                           PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_261                           PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_280                           PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_281                           PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_302                           PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_303                           PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_305                           PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_306                           PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_307                           PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_311                           PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_101                           PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_105                           SKIPPED      CoAPthon3 package not installed
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_1630                          PASSED      
INFO    -                                    net.lwm2m.interop.test_LightweightM2M_1_1_int_1635[None]                    PASSED      
```

All the interoperability tests are passing now. Same with module-tests/content-format tests.

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_message_handling.c:
```diff
@@ -1025,15 +1026,15 @@ static int lwm2m_write_handler_opaque(struct lwm2m_engine_obj_inst *obj_inst,
 					       res_inst->res_inst_id, write_buf, len,
 					       last_pkt_block && last_block, opaque_ctx.len,
-					       msg->in.block_ctx->ctx.current);
+						   opaque_ctx.offset);
```
Nitpicking but there is a whitespace problem. Does not align.

DaveLacerteHQ replied:
Sorry about that, fixed
